v3/lints/cabf_smime_br/lint_registration_scheme_id_matches_subject_country.go

```diff
@@ -0,0 +1,88 @@
+ * permissions and limitations under the License.
+ */
+
+package cabf_smime_br
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+)	
+
+var countryRegex = regexp.MustCompile(`^([A-Z]{3})([A-Z]{2})`)
+
+func init() {
+	lint.RegisterCertificateLint(&lint.CertificateLint{
+		LintMetadata: lint.LintMetadata{
+			Name:          "e_registration_scheme_id_matches_subject_country",
+			Description:   "The country code used in the Registration Scheme identifier SHALL match that of the subject:countryName in the Certificate as specified in Section 7.1.4.2.2",
+			Citation:      "Appendix A A.1",
+			Source:        lint.CABFSMIMEBaselineRequirements,
+			EffectiveDate: util.CABF_SMIME_BRs_1_0_0_Date,
+		},
+		Lint: NewRegistrationSchemeIDMatchesSubjectCountry,
+	})
+}
+
+type registrationSchemeIDMatchesSubjectCountry struct{}
+
+// NewRegistrationSchemeIDMatchesSubjectCountry creates a new linter to enforce SHALL requirements for registration scheme identifiers matching subject:countryName
+func NewRegistrationSchemeIDMatchesSubjectCountry() lint.LintInterface {
+    return &registrationSchemeIDMatchesSubjectCountry{}
+}
+
+// CheckApplies returns true if the provided certificate contains subject:countryName and one-or-more of the following SMIME BR policy identifiers:
+//   - Organization Validated Strict
+// 	 - Organization Validated Multipurpose
+// 	 - Organization Validated Legacy
+//   - Sponsor Validated Strict
+//   - Sponsor Validated Multipurpose
+//   - Sponsor Validated Legacy
+func (l *registrationSchemeIDMatchesSubjectCountry) CheckApplies(c *x509.Certificate) bool {
+	if c.Subject.Country == nil {
+		return false
+	}
+	return util.IsOrganizationValidatedCertificate(c) || util.IsSponsorValidatedCertificate(c)
+}
+
+// Execute applies the requirements on matching subject:countryName with registration scheme identifiers
+func (l *registrationSchemeIDMatchesSubjectCountry) Execute(c *x509.Certificate) *lint.LintResult {
+	country := c.Subject.Country[0]
+	
+	for _, id := range c.Subject.OrganizationIDs {
+		if err := verifySMIMEOrganizationIdentifierContainsSubjectNameCountry(id, country); err != nil {
+			return &lint.LintResult{Status: lint.Error, Details: err.Error()}
+		}
+	}
+	return &lint.LintResult{Status: lint.Pass}
+}
+
+// verifySMIMEOrganizationIdentifierContainSubjectNameCountry verifies that the country code used in the subject:organizationIdentifier matches subject:Name:Country
+func verifySMIMEOrganizationIdentifierContainsSubjectNameCountry(id string, country string) error {
+	submatches := countryRegex.FindStringSubmatch(id)
+	if len(submatches) < 3 {
+		return fmt.Errorf("subject:organizationIdentifier field SHALL contain a Registration Reference for a Legal Entity assigned in accordance to the identified Registration Scheme")
+	}
+
+	identifierCountry := submatches[2]
+
+	if identifierCountry != country {
```

Comment:
This is where I think you might get into difficulties.  What if the OID starts GOVGB and the country in the subject is "Great Britain".  I would argue that by the text of the BRs, the country code in the organization identifier _does_ "match that of the subject:countryName" as the word 'that' is referring to the country code of the subject:countryName rather than suggesting the subject:countryName should contain the ISO country code.

Reply:
Nevermind.  The BRs state that subject:countryName must contain an ISO country code:

Certificate Field: subject:countryName (OID: 2.5.4.6)
Contents: If present, the subject:countryName SHALL contain the two‐letter ISO 3166‐1
country code associated with the location of the Subject verified under Section 3.2.3 for
pg. 65
Organization‐validated and Sponsor‐validated Certificate Types or Section 3.2.4 for
Individual‐validated Certificate Types. If a Country is not represented by an official ISO 3166‐1
country code, the CA MAY specify the ISO 3166‐1 user‐assigned code of XX indicating that an
official ISO 3166‐1 alpha‐2 code has not been assigned.

Reply:
Thanks for checking and confirming this Rob.